File: typescript/src/analysis.ts
/**
 * Geodesic measurement and raster classification operations.
 * @module analysis
 */

import type { RoteiroClient } from './client';
import type {
  ClassifyResult,
  FeatureCollection,
  GeodesicAreaResult,
  GeodesicLengthResult,
} from './types';

/**
 * Compute geodesic polygon area in square meters on the WGS84 ellipsoid.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param geojson - A GeoJSON FeatureCollection containing polygon features.
 * @returns A list of results with index and area_sq_m for each polygon.
 */
export async function geodesicArea(
  client: RoteiroClient,
  geojson: FeatureCollection,
): Promise<GeodesicAreaResult[]> {
  return client.post<GeodesicAreaResult[]>('/api/geodesic/area', geojson);
}

/**
 * Compute geodesic linestring length in meters on the WGS84 ellipsoid.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param geojson - A GeoJSON FeatureCollection containing linestring features.
 * @returns A list of results with index and length_m for each linestring.
 */
export async function geodesicLength(
  client: RoteiroClient,
  geojson: FeatureCollection,
): Promise<GeodesicLengthResult[]> {
  return client.post<GeodesicLengthResult[]>('/api/geodesic/length', geojson);
}

/**
 * K-means unsupervised pixel classification.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param params - Classification parameters with bands and config.
 * @returns A classification result with labels, width, height, and optional confidence.
 */
export async function classifyKMeans(
  client: RoteiroClient,
  params: { bands: Record<string, unknown>; config: { k: number; max_iterations?: number } },
): Promise<ClassifyResult> {
  return client.post<ClassifyResult>('/api/raster/classify/kmeans', params);
}

/**
 * ISODATA unsupervised pixel classification.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param params - Classification parameters with bands and config.
 * @returns A classification result with labels, width, height, and optional confidence.
 */
export async function classifyIsodata(
  client: RoteiroClient,
  params: { bands: Record<string, unknown>; config: Record<string, unknown> },
): Promise<ClassifyResult> {
  return client.post<ClassifyResult>('/api/raster/classify/isodata', params);
}

/**
 * Maximum likelihood supervised pixel classification.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param params - Classification parameters with bands and training data.
 * @returns A classification result with labels, width, height, and optional confidence.
 */
export async function classifyML(
  client: RoteiroClient,
  params: { bands: Record<string, unknown>; training: Record<string, unknown>[] },
): Promise<ClassifyResult> {
  return client.post<ClassifyResult>('/api/raster/classify/ml', params);
}

/**
 * Random forest supervised pixel classification.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param params - Classification parameters with bands, training data, and optional config.
 * @returns A classification result with labels, width, height, and optional confidence.
 */
export async function classifyRF(
  client: RoteiroClient,
  params: { bands: Record<string, unknown>; training: Record<string, unknown>[]; config?: Record<string, unknown> },
): Promise<ClassifyResult> {
  return client.post<ClassifyResult>('/api/raster/classify/rf', params);
}

File: python/roteiro/analysis.py
"""Geodesic measurement and raster classification operations."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from .client import RoteiroClient


def geodesic_area(client: RoteiroClient, geojson: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Compute geodesic polygon area in square meters on the WGS84 ellipsoid.

    Args:
        client: An initialised RoteiroClient instance.
        geojson: A GeoJSON FeatureCollection containing polygon features.

    Returns:
        A list of results with ``index`` and ``area_sq_m`` for each polygon.
    """
    return client._post("/api/geodesic/area", geojson)


def geodesic_length(client: RoteiroClient, geojson: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Compute geodesic linestring length in meters on the WGS84 ellipsoid.

    Args:
        client: An initialised RoteiroClient instance.
        geojson: A GeoJSON FeatureCollection containing linestring features.

    Returns:
        A list of results with ``index`` and ``length_m`` for each linestring.
    """
    return client._post("/api/geodesic/length", geojson)


def classify_kmeans(
    client: RoteiroClient,
    bands: Dict[str, Any],
    config: Dict[str, Any],
) -> Dict[str, Any]:
    """K-means unsupervised pixel classification.

    Args:
        client: An initialised RoteiroClient instance.
        bands: Band data for classification.
        config: Configuration with ``k`` and optional ``max_iterations``.

    Returns:
        A classification result with ``labels``, ``width``, ``height``,
        and optional ``confidence``.
    """
    return client._post("/api/raster/classify/kmeans", {"bands": bands, "config": config})


def classify_isodata(
    client: RoteiroClient,
    bands: Dict[str, Any],
    config: Dict[str, Any],
) -> Dict[str, Any]:
    """ISODATA unsupervised pixel classification.

    Args:
        client: An initialised RoteiroClient instance.
        bands: Band data for classification.
        config: ISODATA configuration parameters.

    Returns:
        A classification result with ``labels``, ``width``, ``height``,
        and optional ``confidence``.
    """
    return client._post("/api/raster/classify/isodata", {"bands": bands, "config": config})


def classify_ml(
    client: RoteiroClient,
    bands: Dict[str, Any],
    training: List[Any],
) -> Dict[str, Any]:
    """Maximum likelihood supervised pixel classification.

    Args:
        client: An initialised RoteiroClient instance.
        bands: Band data for classification.
        training: Training sample data for supervised classification.

    Returns:
        A classification result with ``labels``, ``width``, ``height``,
        and optional ``confidence``.
    """
    return client._post("/api/raster/classify/ml", {"bands": bands, "training": training})


def classify_rf(
    client: RoteiroClient,
    bands: Dict[str, Any],
    training: List[Any],
    config: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Random forest supervised pixel classification.

    Args:
        client: An initialised RoteiroClient instance.
        bands: Band data for classification.
        training: Training sample data for supervised classification.
        config: Optional random forest configuration parameters.

    Returns:
        A classification result with ``labels``, ``width``, ``height``,
        and optional ``confidence``.
    """
    body: Dict[str, Any] = {"bands": bands, "training": training}
    if config:
        body["config"] = config
    return client._post("/api/raster/classify/rf", body)
